### DISCUSSION TITLE
What does #(false) mean, and how can I detect that a node of this type is being built?

### DISCUSSION THREAD

**Discussion by vsajip** [javacc21/javacc21#224] [Q&A]
https://github.com/javacc21/javacc21/blob/0637b2cf7443914ee009320703f2a72f09549a06/examples/java/Java.javacc#L84

**Comment by revusky** [comment]
See: https://javacc.com/2021/01/12/just-like-home-made/

It means that we never create the Node for this production but we still generate a `Modifiers` class. I can't tell you offhand why it isn't just:

            Modifiers#void :

The difference is that if it is annotated with `#void` it does not generate the class. Well, the answer is that, for some reason, at some point, I wanted to generate a `Modifiers` class and presumably use it, but then I guess I didn't.

**Comment by vsajip** [comment]
OK, thanks for the pointer. ISTM a `Modifiers` class would be useful from a tree-building point of view to encapsulate all the modifiers. As it stands, the modifiers are just collected in a list and not separated out from other stuff in a declaration, making it more awkward than necessary to answer questions at a high level about modifiers for a particular declaration. Of course, that's the kind of thing I need when transpiling :slightly_smiling_face:

**Comment by vsajip** [comment]
Hmmm. After changing `Modifiers#(false) :` to `Modifiers :`, updating the bootstrap jar and running `ant test`, a "Code too large" error is hit when parsing the C# grammar:
```
compile:
    [javac] Compiling 250 source files
    [javac] /disk2/projects/javacc21/examples/csharp/org/parsers/csharp/CSharpParser.java:399: error: code too large
    [javac]     private static HashMap<TokenType[], EnumSet<TokenType> > enumSetCache= new HashMap<> ();
    [javac]                                                              ^
    [javac] 1 error
```
I can't see what exactly could be causing this error, perhaps you can? I attach the source file.
[CSharpParser.java.txt](https://github.com/javacc21/javacc21/files/9433770/CSharpParser.java.txt)

**Reply by revusky** [reply]
I think I see the problem. Turning the Modifiers into a Node ended up breaking the [DeadCodeEliminator](https://github.com/javacc21/javacc21/blob/master/src/java/com/javacc/output/java/DeadCodeEliminator.java) logic which was identifying private methods and fields a certain way. For example, look at line 78 of DeadCodeEliminator.java and you see that the change breaks it because it now would need to get the Modifiers node and see if it contains a PRIVATE token. You see?

And with DeadCodeEliminator broken, we can hit "Code too large" because the way the code works is that it generates a lot of stuff that is ultimately unneeded and then uses the DeadCodeEliminator logic to strip it out. But with that broken, it doesn't get stripped out and we can hit "Code too large". 

You see? Isn't that nastily tricky?

But you can fix it pretty easily! Actually DeadCodeEliminator is pretty crude and you might feel compelled to make it more correct, but it might not be a good use of your time...

**Comment by vsajip** [comment]
> You see? Isn't that nastily tricky?

It certainly is! I wouldn't have thought of looking there ...

**Comment by vsajip** [comment]
> the way the code works is that it generates a lot of stuff that is ultimately unneeded and then uses the DeadCodeEliminator logic to strip it out.

Better would be some means of not generating the code at all, if we know it isn't going to be used.  The way it is now, it's pretty brittle (plus, it doesn't benefit the other generation languages at all).

**Reply by revusky** [reply]
>Better would be some means of not generating the code at all, if we know it isn't going to be used.

Well, it's not all that clear that it's better. Simply generating a whole bunch of scanahead code that may not be used and then stripping it out in a later step is actually a pretty good approach and I don't really see offhand why it is more "brittle" than alternatives. 

For one thing, if you look at DeadCodeEliminator.java, that whole class is not even 100 lines of code. In fact, I already patched it so that it can handle the modifiers being in a Modifiers holder. Though, it also turned out that I had to patch some other places in the code. For example, in JavaCodeUtils.java, [here](https://github.com/javacc21/javacc21/blob/master/src/java/com/javacc/output/java/JavaCodeUtils.java#L144) notably, I have to take into account that an Annotation could be in the Modifiers holder class. So you see the little patch necessary there. (You realize, I assume, that an Annotation is also a "modifier"). Or [here](https://github.com/javacc21/javacc21/blob/master/src/java/com/javacc/output/java/JavaCodeUtils.java#L60) you see that I needed to patch the logic a bit to take into account that the Annotations could be in a Modifiers holder.

So, okay, granted, if you change the shape of the tree, then things can break and one needs to adjust. But the full set of diffs necessary to handle this was actually quite small. And the advantage of using the tree internally is that actually you end up with something more robust because it's an extra way the system is getting stressed.

**Comment by vsajip** [comment]
I set breakpoints in `DeadCodeEliminator` on method and field declaration removal and ran JavaCC on the C# grammar. I only saw a few small methods and a few fields get excluded - see what I have commented out/marked for future conditional generation in #225. Did I miss some others? I set breakpoints here:
https://github.com/javacc21/javacc21/blob/d2274d3514c91ca8b57084ed4930c2eb406cea70/src/java/com/javacc/output/java/DeadCodeEliminator.java#L67
https://github.com/javacc21/javacc21/blob/d2274d3514c91ca8b57084ed4930c2eb406cea70/src/java/com/javacc/output/java/DeadCodeEliminator.java#L131
https://github.com/javacc21/javacc21/blob/d2274d3514c91ca8b57084ed4930c2eb406cea70/src/java/com/javacc/output/java/DeadCodeEliminator.java#L134

and after doing `update-bootstrap`, even in the branch representing #225 (`Modifiers` production unchanged) I'm now getting code too large. Not sure if this an issue just under the surface for the C# grammar?

**Reply by revusky** [reply]
Well, the main thing that had to get patched in DeadCodeEliminator.java was the logic for determining whether a field or method is private. [The method here](https://github.com/javacc21/javacc21/blob/master/src/java/com/javacc/output/java/DeadCodeEliminator.java#L76) now works whether the "private" token is at the top level or inside a Modifiers holder. Since the dead code eliminator only eliminates methods or fields that are private (because otherwise we don't know whether they are being used) if you're not identifying what is private, then you're just not eliminating anything.

**Comment by vsajip** [comment]
> I'm now getting code too large

Ignore that, it was a build problem.

**Comment by vsajip** [comment]
I'm now seeing things like `ClassDeclaration_FIRST_SET_init()` which I wasn't before, I think because the Java file being looked at had already had them stripped from an earlier javacc run. I didn't clean the generated source code :disappointed:
